### PR TITLE
Thomas datadownloader

### PR DESCRIPTION
**Pull request by t6george**
The data downloader part of the pipeline is done. So when training our models, we can download the (image, natural language) pairs we have collected to the local machine for easy access. The usage for the command line arguments is described if you run with the -h flag (can directly view the command descriptions in arg_parser.py).

Also in honour of Peter Buhr, I made this thing (optionally) multi-threaded so that it can take advantage of Leo's 16-CPU computer that we will eventually train on.

### CODE REVIEW

**Comment by q349wang** [comment]
O only have a 12 core CPU

**Comment by t6george** [comment]
> I only have a 12 core CPU

File: pipeline/datadownloader/dataset_downloader.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                threading.Thread(target=file_writer,
+                                 args=(i, num_extra_threads + 1, csv_data,
+                                       base_dir))
+                for i in range(2, num_extra_threads + 2)
+            ]
+
+            for thread in threads:
+                thread.start()
+
+            file_writer(1, num_extra_threads + 1, csv_data, base_dir)
```
Comment (q349wang) [inline pipeline/datadownloader/dataset_downloader.py:67]: Is this meant to be 1-based indexing?

File: pipeline/datadownloader/file_writer.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+class Attributes(Enum):
+    IMG_NAME = 0,
+    IMG_URL = 1,
+    DESRIPTION = 2
+
+
+def file_writer(thread_id: int, stride: int, csv_data: [[str]], base_dir: str):
+    for row in range(thread_id, len(csv_data), stride):
+        response = requests.get(csv_data[row][1])
+        with open(base_dir + csv_data[row][0] + '.png', 'wb') as imgFile:
```
Comment (q349wang) [inline pipeline/datadownloader/file_writer.py:15]: Could have race condition here. if two files have the same name. Idk if thats possible.

Comment (t6george) [reply pipeline/datadownloader/file_writer.py:15]: No, that would be the case if we downloaded the same image twice. If that happens, it is a problem in the scraper.

File: pipeline/datadownloader/dataset_downloader.py (hunk 1 above)
Comment (t6george) [reply pipeline/datadownloader/dataset_downloader.py:67]: Yes, index 0 corresponds to the column titles of the csv

**Review by q349wang — APPROVED** [review]

**Pull request merged**